server/src/main/java/org/candlepin/policy/ValidationResult.java

```diff
@@ -97,6 +117,14 @@ public boolean hasErrors() {
         return !errors.isEmpty();
```

Comment:
This likely needs to return ```!errors.isEmpty() || !errorKeys.isEmpty();``` now, with another method to check for error strings. Though, I suppose if ```hasErrors``` returns true and ```hasErrorKeys``` returns false, we know that it has to be error strings.

Reply:
I actually envision the `errors` and `warnings` fields to eventually get removed from `ValidationResult` altogether, and errorKeys/warningKeys to be what stays on this class. The only reason they're not removed for now is because other rules use them. So I don't think we should mix logic of the two.

Reply:
I've kept that separate for now (since ActivationKeyRules is only using the errorKeys, and other rules are only using the errors/warnings). If you feel strongly about not having that setup like this, I can create a new class that holds the enums (instead of highjacking ValidationResult).

Reply:
I wonder if the future of the enum version is where the enum has a severity value associated with it, or a boolean to differentiate between warning/error. Or just a separate collection.

Reply:
I'd prefer separate collections, instead of always having to do validationError.isWarning() or validationError.getSeverity() on each of the items in the collection. They're used differently, so they should have different fields/methods.